Edit README up until Clean Go. Fix miscellaneous issues. (#11)
Pungyeon/clean-go-article: README.md

In Go, according to `gofmt`, <em>all</em> public variables and functions should be annotated. I think this is absolutely fine, as it gives us consistent rules for documenting our code. However, I always want to distinguish between comments that enable auto-generated documentation and <em>all other</em> comments. Annotation comments, for documentation, should be written like documentation&mdash;they should be at a high level of abstraction and concern the logical implementation of the code as little as possible.

> I say this because there are other ways to explain code and ensure that it's being written comprehensibly and expressively. If the code is neither of those, some people find it acceptable to introduce a comment explaining the convoluted logic. Most people simply won't read comments because they tend to be very intrusive to the experience of reviewing code.

Reviewer: ```suggestion
I say this because there are other ways to explain code and ensure that it's being written comprehensibly and expressively. If the code is neither of those, some people find it acceptable to introduce a comment explaining the convoluted logic. Most people simply won't read comments, as they tend to be very intrusive to the experience of reviewing code.
```

}
```

> Here, we've clearly distinguished the nested function calls from their parent without being overly specific. This allows each nested function call to make sense on its own as well as within the context of the parent. On the other hand, if we had named the `parseJSON` function `json` instead, it couldn't possibly stand on its own. What about JSON? Are we parsing it? Creating a JSON object? Hopefully you get the point.

Reviewer: I think I would prefer a slight adjustment on the last sentence, to keep it formal. What do you say?

```suggestion
Here, we've clearly distinguished the nested function calls from their parent without being overly specific. This allows each nested function call to make sense on its own as well as within the context of the parent. On the other hand, if we had named the `parseJSON` function `json` instead, it couldn't possibly stand on its own. The functionality would become lost in the name, and we are no longer able to tell whether this function is parsing, creating or marshalling json. 
```

Author: Agreed! This definitely reads better. I'd slightly reword the part after the comma:

", and we would no longer be able to tell..."

And then also this:

"creating[comma] or marshalling JSON."

(I was unable to add a suggestion to the batch to make things easier.)

Reviewer: That makes perfect sense! Let's go with that.

Notice that `fileExtension` is actually a little more specific. However, this is because its functionality is in fact quite specific in nature:

```go
func fileExtension(filepath string) string {
    segments := strings.Split(filepath, ".")
    return segments[len(segments)-1]
}
```

This kind of logical progression in our function names&mdash;from a high level of abstraction to a lower, more specific one&mdashmakes the code easier to follow and and read. Consider the alternative: If our highest level of abstraction is too specific, then we'll end up with a name that attempts to cover all bases, like `DetermineFileExtensionAndParseConfigurationFile`. This is horrendously difficult to read; we are trying to be too specific too soon and end up confusing the reader, despite trying to be clear! 

#### Variable Naming
Rather interestingly, the opposite is true for variables. Unlike functions, our variables should be named from more to less specific the deeper we go into nested scopes.

> <em>You shouldn’t name your variables after their types for the same reason you wouldn’t name your pets 'dog' or 'cat'. &ndash; Dave Cheney</em>

> Why should our variable names become less specific as we travel deeper into a function's scope? Simply put, as a variable's scope becomes smaller, it becomes increasingly clear for the reader what that variable represents, so there's no need to use a specific name. In the example of the previous function `fileExtension`, we could even shorten the name of the variable `segments` to `s` if we wanted to. The context of the variable is so clear that it's unnecessary to explain it any further with longer variable names. Another good example of this is in nested for loops:

Reviewer: ```suggestion
Why should our variable names become less specific as we travel deeper into a function's scope? Simply put, as a variable's scope becomes smaller, it becomes increasingly clear for the reader what that variable represents and thereby eliminates the need for specific naming. In the example of the previous function `fileExtension`, we could even shorten the name of the variable `segments` to `s` if we wanted to. The context of the variable is so clear that it's unnecessary to explain it any further with longer variable names. Another good example of this is in nested for loops:
```

Author: Works for me! Though I'd maybe write it this way to be technically correct:

"...represents, thereby eliminating the need..."

(Unless you wanted to avoid using too many commas.)

Reviewer: no, I agree that this is better 👍

    - get order from database
```

By writing short functions (which are typically 5&ndash;8 lines in Go), we can create code that reads almost as naturally as our description above:

```go
var (
    NullItem = Item{}

>     ErrInsufficientPrivileges = errors.New("user does not have sufficient privileges")

Reviewer: 😅

Author: I wouldn't have even caught it if I hadn't copy-pasted my edited parts in Word, lol.

)
```

> The function `QueueDeclare` takes six input parameters, which is quite a lot. It's possible to understand what this code does with some effort because of the comments. And therein lies the problem! As mentioned earlier, comments should be substituted with descriptive code. After all, there's nothing preventing us from invoking the `QueueDeclare` function without comments, making it look like this:

Reviewer: ```suggestion
The function `QueueDeclare` takes six input parameters, which is quite a lot. With some effort though, it's possible to understand what this code does, because of the comments. But, therein lies the problem! As mentioned earlier, comments should be substituted with descriptive code. After all, there's nothing preventing us from invoking the `QueueDeclare` function without comments, making it look like this:
```

Reviewer: Honestly, not that happy with my suggestion either... but I feel the sentence is a little clumsy... Any ideas?

Author: Yeah, I agree! It's worded a bit awkwardly.

How about this? I'm open to your suggestions here as well.

> The function `QueueDeclare` takes six input parameters, which is quite a lot. With some effort, it's possible to understand what this code does thanks to the comments. However, the comments are actually part of the problem&mdash;as mentioned earlier, they should be substituted with descriptive code whenever possible. After all, there's nothing preventing us from invoking the `QueueDeclare` function <em>without</em> comments:

Reviewer: Nice, this is much more elegant.

}
```

> However, we can do even better by invoking the function directly after it's declared. This makes it much clearer that the function logic is associated with the declared variable:

Reviewer: ```suggestion
However, we can do even better by invoking the function directly after it's declaration. This makes it much clearer that the function logic is associated with the declared variable:
```

Author: That works too :) We'd just need to get rid of the apostrophe from "it's" for it to be correct.

Reviewer: ah, shit 😅 nice catch.
